Signals.cpp implementation for Windows

iangrunert opened the pull request:
#### 66b9bc907a2e984b5f6c15729edad082f0a59a83
<pre>
Signals.cpp implementation for Windows
<a href="https://bugs.webkit.org/show_bug.cgi?id=259108">https://bugs.webkit.org/show_bug.cgi?id=259108</a>

Reviewed by Yusuke Suzuki.

Signals is implemented using a VectoredExceptionHandler on Windows, so
we cannot support all the Signal types as the other OS&apos;s. We need
support for AccessFault, which is used for WASM memory.

Test: Tools/TestWebKitAPI/Tests/WTF/Signals.cpp

* Source/JavaScriptCore/wasm/WasmFaultSignalHandler.cpp:
* Source/JavaScriptCore/wasm/WasmFaultSignalHandler.h:
* Source/WTF/wtf/PlatformJSCOnly.cmake:
* Source/WTF/wtf/PlatformWin.cmake:
* Source/WTF/wtf/WTFConfig.h:
* Source/WTF/wtf/threads/Signals.cpp:
(WTF::toSystemSignal):
(WTF::fromSystemSignal):
* Source/WTF/wtf/threads/Signals.h:
(WTF::toSystemSignal): Deleted.
(WTF::fromSystemSignal): Deleted.
* Source/WTF/wtf/win/SignalsWin.cpp: Added.
(WTF::SignalHandlers::add):
(WTF::SignalHandlers::forEachHandler const):
(WTF::fromSystemException):
(WTF::vectoredHandler):
(WTF::addSignalHandler):
(WTF::activateSignalHandlersFor):
(WTF::SignalHandlers::initialize):
* Tools/TestWebKitAPI/CMakeLists.txt:
* Tools/TestWebKitAPI/Tests/WTF/Signals.cpp:
(TEST):

Canonical link: <a href="https://commits.webkit.org/266716@main">https://commits.webkit.org/266716@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8c0cb3f77cc8214203839b0b6a17fc73666de145

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/14541 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/14852 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/15196 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/16288 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/13747 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/14677 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/17367 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/14929 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/16402 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/14723 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/15244 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/12350 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/17017 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/12527 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/13116 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/20128 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/12451 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/13606 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/13281 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/16523 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/13827 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/13830 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/11658 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/14586 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/13122 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/3775 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/3516 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/17459 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/14975 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/13674 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/3589 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/WTF/wtf/threads/Signals.h:
```diff
@@ -174,16 +175,93 @@ using WTF::SignalHandler;
+// and once commited they can't be turned off.
+WTF_EXPORT_PRIVATE void addSignalHandler(Signal, SignalHandler&&);
+WTF_EXPORT_PRIVATE void activateSignalHandlersFor(Signal);
+
```
A lot of declarations are duplicate to UNIX ones. Can you share and dedup them?

Constellation reviewed (changes requested):
Nice!

donny-dont commented on Source/WTF/wtf/WTFConfig.h:
```diff
@@ -83,7 +83,7 @@ struct Config {
     int sigThreadSuspendResume;
 #endif
-#if OS(UNIX)
+#if OS(UNIX) || OS(WINDOWS)
```
@Constellation is there any `OS` that isn't `UNIX` or `WINDOWS`? Only one I'm not sure about is `FUSCIA`.

If not then the guard can just be dropped.

iangrunert replied:
https://github.com/WebKit/WebKit/blob/main/Source/WTF/wtf/PlatformOS.h#L121-L134

`FUCHSIA` gets bundled as Unix-like, so I think we could drop this guard.

donny-dont commented on Source/WTF/wtf/threads/Signals.h:
```diff
@@ -174,16 +175,93 @@ using WTF::SignalHandler;
 using WTF::activateSignalHandlersFor;
 
-#else // not OS(UNIX)
+#elif OS(WINDOWS) // not OS(UNIX)
```
If there is no `OS` that is neither `UNIX` or `WINDOWS` then this can just be an `#else`.

donny-dont commented on Source/WTF/wtf/threads/Signals.h:
```diff
@@ -174,16 +175,93 @@ using WTF::SignalHandler;
+    default:
+        return Signal::Unknown;
+    }
+}
```
I know this is mirroring `fromSystemSignal` but looking at the codebase it seems like both it and `toSystemSignal` aren't used outside of the `Signals.cpp` file so they both can be moved in there.

For the windows bits it saves us an include of `windows.h` if you just move it into the .cpp file

donny-dont commented on Source/WTF/wtf/threads/Signals.h:
```diff
@@ -56,8 +59,16 @@ enum class Signal {
+    AccessFault,
+    NumberOfSignals = AccessFault + 1,
+    Unknown
+#endif
```
@Constellation do you want these shared as well? I would say yes.

Constellation replied:
Yeah, that's nice if we can share them. But there are a bit of hacks like `NumberOfSignals = AccessFault + 2`, so it is OK to keep them separate.

donny-dont commented on Source/WTF/wtf/threads/Signals.h:
```diff
@@ -89,6 +100,7 @@ inline Signal fromSystemSignal(int signal)
     default: return Signal::Unknown;
     }
 }
+#endif // OS(UNIX)
```
This could be a follow up but I would move these into the `Signals.cpp` file. My grepping showed that neither function was used outside of `Signals.cpp`.

Constellation replied:
That sounds good.

donny-dont reviewed (commented):
LGTM minus some nits. Will let @Constellation have the final word about the shared enum.

Constellation commented on Source/WTF/wtf/threads/Signals.h:
```diff
@@ -56,8 +59,16 @@ enum class Signal {
+    IllegalInstruction,
+    AccessFault,
+    NumberOfSignals = AccessFault + 1,
+    Unknown
```
We should make `Unknown = NumberOfSignals` since NumberOfSignals is not a signal.

Constellation reviewed (approved):
r=me with comments.

donny-dont commented on Source/WTF/wtf/threads/Signals.cpp:
```diff
@@ -336,6 +335,38 @@ void registerThreadForMachExceptionHandling(Thread& thread)
 
 #endif // HAVE(MACH_EXCEPTIONS)
 
+inline std::tuple<int, std::optional<int>> toSystemSignal(Signal signal)
+{
+    switch (signal) {
+    case Signal::AccessFault: return std::make_tuple(SIGSEGV, SIGBUS);
+    case Signal::IllegalInstruction: return std::make_tuple(SIGILL, std::nullopt);
+    case Signal::Usr: return std::make_tuple(SIGUSR2, std::nullopt);
+    case Signal::FloatingPoint: return std::make_tuple(SIGFPE, std::nullopt);
+    case Signal::Breakpoint: return std::make_tuple(SIGTRAP, std::nullopt);
+#if !OS(DARWIN)
+    case Signal::Abort: return std::make_tuple(SIGABRT, std::nullopt);
+#endif
+    default: break;
+    }
+    RELEASE_ASSERT_NOT_REACHED();
+}
+
+inline Signal fromSystemSignal(int signal)
+{
+    switch (signal) {
+    case SIGSEGV: return Signal::AccessFault;
+    case SIGBUS: return Signal::AccessFault;
+    case SIGFPE: return Signal::FloatingPoint;
+    case SIGTRAP: return Signal::Breakpoint;
+    case SIGILL: return Signal::IllegalInstruction;
+    case SIGUSR2: return Signal::Usr;
+#if !OS(DARWIN)
+    case SIGABRT: return Signal::Abort;
+#endif
+    default: return Signal::Unknown;
+    }
+}
+
```
Put this at the top of the block and the build should go back to green.

donny-dont commented on Source/WTF/wtf/threads/Signals.h:
```diff
@@ -56,39 +59,17 @@ enum class Signal {
-    }
-}
+#if OS(UNIX)
+#endif // OS(UNIX)
```
Remove this its empty.

donny-dont reviewed (commented):
Almost there!

iangrunert committed:
Signals.cpp implementation for Windows
https://bugs.webkit.org/show_bug.cgi?id=259108

Reviewed by Yusuke Suzuki.

Signals is implemented using a VectoredExceptionHandler on Windows, so
we cannot support all the Signal types as the other OS's. We need
support for AccessFault, which is used for WASM memory.

Test: Tools/TestWebKitAPI/Tests/WTF/Signals.cpp

* Source/JavaScriptCore/wasm/WasmFaultSignalHandler.cpp:
* Source/JavaScriptCore/wasm/WasmFaultSignalHandler.h:
* Source/WTF/wtf/PlatformJSCOnly.cmake:
* Source/WTF/wtf/PlatformWin.cmake:
* Source/WTF/wtf/WTFConfig.h:
* Source/WTF/wtf/threads/Signals.cpp:
(WTF::toSystemSignal):
(WTF::fromSystemSignal):
* Source/WTF/wtf/threads/Signals.h:
(WTF::toSystemSignal): Deleted.
(WTF::fromSystemSignal): Deleted.
* Source/WTF/wtf/win/SignalsWin.cpp: Added.
(WTF::SignalHandlers::add):
(WTF::SignalHandlers::forEachHandler const):
(WTF::fromSystemException):
(WTF::vectoredHandler):
(WTF::addSignalHandler):
(WTF::activateSignalHandlersFor):
(WTF::SignalHandlers::initialize):
* Tools/TestWebKitAPI/CMakeLists.txt:
* Tools/TestWebKitAPI/Tests/WTF/Signals.cpp:
(TEST):

Canonical link: https://commits.webkit.org/266716@main
